Issue management documentation (#85)
IMAP-Science-Operations-Center/imap_processing: docs/source/development/style-guide/requirement-and-issue-tracking.rst

> .. _requirement-and-issue-tracking:

Reviewer: This document is great. But, I don't think it belongs under the style guide. I think it should go under a new subfolder - maybe `development/project-management`? or just `source/project-management`, for all github issues documentation and such? Actually, the PR document should probably go under that folder as well... what do you think?

Author: Yeah, that makes sense. I'll create a project management subfolder. I'm planning to add a page for a SIT timeline with mini-milestones to prepare for each SIT, which could also make sense under project management.

Level 4 requirement requirements that have been decomposed to the SDC level. Level 4 requirements have already been defined.
Level 5 requirement requirements decomposed from the level 4 requirements that better define how the SDC plans to satisfy the parent level 4 requirement, while not defining any implementation details. Example: Backup *Metadata* to AWS backup account vs. Backup *OpenSearch* to AWS backup account
Implementation task Github issue that better defines how a certain feature will be implemented. Implementation tasks are tracked under the parent level 5 requirement as a sub-task. Each implementation task should be small enough to be handled in a single PR.
SPIKE               Github issue created to investigate different architecture design approaches. Like implementation tasks, SPIKEs should be tracked under the applicable level 5 requirement as a sub-task.
=================== ===================================================================================================================================================================================================================================================

Requirements Overview
^^^^^^^^^^^^^^^^^^^^^

> Requriements Mapping

Reviewer: ```suggestion
Requirements Mapping
```

~~~~~~~~~~~~~~~~~~~~

* Each level 4 requirement is mapped to at least one SIT.

> * Each level 4 requirement relevant to the upcoming SIT is decomposed into level 5 requirements that, together, cover the testing requirements of the SIT.

Reviewer: ```suggestion
* Each level 4 requirement within an upcoming SIT is split into level 5 requirements. These level 5 requirements, taken together, should completely cover the level 4 requirements. 
```

This bullet point was a little jargon-y and I had trouble understanding what you meant. Let me know if my edit is correct!

Author: I like that a lot better! The only thing I would change is "Each level 4 requirement being tested in the upcoming SIT" or "Each level 4 requirement mapped to the upcoming SIT". I just don't want to say "within" because level 4s can span multiple SITs. Do you have a preference?

Reviewer: I think your change makes sense!

> * The level 5 requirement creation is incremental as we move through development for each SIT.

Reviewer: ```suggestion
* The level 5 requirements are created as we move through development for each SIT.
```

* Each Level 5 requirement will have implementation tasks (sub-issues) that fully cover the Level 5 requirement and contain implementation specific information.

> .. image:: ../../_static/req-decomposition.png

Reviewer: The images are great for making things more clear!

   :alt: Requirement decomposition diagram
   :width: 1293
   :height: 629

Requirements Lifecycle
~~~~~~~~~~~~~~~~~~~~~~
The below image shows an example of the lifecycle of a Level 4 requirement over the course of the project:

.. image:: ../../_static/req-lifecycle.png

>    :alt: Requirement lifecycle diagram

Reviewer: This diagram is quite squished on my screen. Maybe because I zoomed in to make the text bigger? Not sure how to fix it, but maybe you can look into a more responsive size option (like making it take the full width of the screen, or something)

Author: Yeah, I guess if there's a way to have the left index and right table of contents closer to the end of the window that would create more space for the images. I can look into if there's a way to do that.

   :width: 1631
   :height: 621

Overview of Managing Requirements in Github
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> * All level 5 requirements will be managed as Github issues.

Reviewer: Link to our github board/issues

Author: I'll add that link

* The status of the requirement issue ("Open" / "Closed") is an indication of both development and testing status, so a requirement is not "Closed" until it is implemented and has passed a SIT.

> * Requirement issues should track sub-issues that describe the implementation.

Reviewer: How should they be tracked? With a checklist of issues is how we've done it before, but would be good to call it out. Those checklists also provide a visual indicator of how "finished" a task is, so we can see if the requirement is complete even without relying on open/closed.

Reviewer: I see this is already covered in the "creating level 5 requirements" section, so I am ok with leaving this as-is

   * Unlike the L5 requirement issues, these issues close once the code is merged.
   * Sub-issues allow for better tracking of how SPIKES, implementation issues, and requirements are related to each other.  They also allow the use of L5 requirements as a starting point to see the design history of each feature.

> * If a re-design is proposed that impacts an already tested requirement, that requirement issue should be re-opened and any new SPIKEs or implementation tasks should be appended to the original task list.

Reviewer: Also, we should link to or summarize whatever discussion led to us re-designing that requirement.

Author: Yeah, I can add more info in the text here. In the requirement issue lifecycle diagram, I said a comment needs to be added when the requirement is re-opened explaining what design change is being proposed.

   * Doing this will keep a record of the evolution of implementation decisions that impacted the requirement and a record of the requirement status throughout the project.
   * This is useful for providing context to new team members or anyone who needs to remember why a particular design was chosen to satisfy a requirement vs another approach.
* A backlog grooming will take place once per month to re-prioritize and clean up the issue board.

Level 5 Requirements in Github
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Creating level 5 requirements
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
* Level 5 requirements should be created using the "L5 Requirement" issue template.
   * As part of this template, the L5 summary, Parent (Level 4) Requirement, and list of tasks (sub-issues) are filled in.
   * If you don't have sub-issues created to link in the task list, you can add text as a reminder or add them later. 
   * The task list should only be appended to for the lifetime of the project.

> * Upon creation a requirement should be given the following the tags:

Reviewer: ```suggestion
* Upon creation, a requirement should be given the following tags:
```

   * "Requirement: Level 5" (you get this for free by selecting the "L5 Requirement" issue template)
   * The appropriate parent requirement category label (Ex. Parent Req: SDC Data Transfer Monitoring)
   * "Untested" (This describes the testing status of the requirement)
* Each requirement should be added to the IMAP project board.
* Each requirement should be assigned a SIT milestone.

Example:

.. image:: ../../_static/l5-label-example.png
   :alt: github requirement issue label example
   :width: 398
   :height: 448

Github Requirement Issue Lifecycle
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The following diagram gives an example of the general information a level 5 requirement issue should contain as it evolves over the lifecycle of development and testing:

> 

Reviewer: That diagram is useful, but very text-heavy, and I find it a little difficult to keep track of what it's saying. Could you simplify the diagram and make the text bigger?

Author: Yeah, I'll try to make this easier to follow. I wanted to have one diagram that people can look at to see what happens to a requirement issue at each stage of development and test, but I agree, I'll try to simplify and reduce the text

.. image:: ../../_static/git-req-issue-lifecycle.png
   :alt: Git Requirement issue lifecycle diagram
   :width: 2161
   :height: 901

> Adding and Tracking sub-issues

Reviewer: This is a great section. So great, that it might be useful in a different page which can be linked to. For example, in the PR documentation, I would link out to this. Could we move this and the other issue creation sections below into another document (maybe "Github Issue Creation")? Ideally that page would include other standards for issue creation eventually, but for now, you can leave that part off and we can add on later.

Author: Yeah, I like this idea!
